### PR TITLE
feat: enhance parser with concurrency constructs

### PR DESCRIPTION
**Pull request by alvynben**
**Summary**

This PR introduces new parsing capabilities to the Go parser, focusing on concurrency constructs and enhancing the language's support for channels and goroutines. Key changes include:

- Added `SelectStatement`, `GoStatement`, and `SendStatement` to support Go's concurrency patterns.
- Included the `<-` unary operator for channel operations.

**Testing**

Usable Code Block
```golang


var a, b = 3,2

func hello() {
	a <- 3
}

func main() {
    hello()
    a = b
    b
    select {
    case a <- 3:
    	a = 4
    case b = <-c:
    	b = 3
      default:
      	a = 3
    }
}

```

### CODE REVIEW

File: src/go-slang/parser/go.pegjs [hunk 1]
```diff
@@ -185,6 +189,46 @@ CallExpression
...
+CommClause
+    = pred:CommCase __ ":" _ statements:Statement* EOS {
+        return { type: "CommClause", pred, statements }
+      }
+CommCase
+    = CASE_TOKEN _ statement:(ReceiveStatement / SendStatement) EOS { return statement}
+    / DEFAULT_TOKEN { return [] }
+
+ReceiveStatement
+    = left:( ExpressionList __ "=" / IdentifierList __ ":=" ) __ right:ReceiveExpression EOS {
```
Comment (shenyih0ng) [inline src/go-slang/parser/go.pegjs]: `left` should be optional according to the spec: https://go.dev/ref/spec#RecvStmt. `case <-ch` should be valid syntax

File: src/go-slang/parser/go.pegjs [hunk 2]
```diff
@@ -185,6 +189,46 @@ CallExpression
...
+    = SELECT_TOKEN _ "{" _ cases:CommClause* _ "}" EOS {
+        return { type: "SelectStatement", cases }
+    }
+
+CommClause
+    = pred:CommCase __ ":" _ statements:Statement* EOS {
+        return { type: "CommClause", pred, statements }
+      }
+CommCase
+    = CASE_TOKEN _ statement:(ReceiveStatement / SendStatement) EOS { return statement}
```
Comment (shenyih0ng) [inline src/go-slang/parser/go.pegjs]:
```suggestion
    = CASE_TOKEN _ statement:(ReceiveStatement / SendStatement) EOS { return statement }
```

File: src/go-slang/parser/go.pegjs [hunk 3]
```diff
@@ -185,6 +189,46 @@ CallExpression
         return makeNode({ type: "CallExpression", callee, args: args ?? [] })
       }
 
+/* Select Statement */
+
+SelectStatement "select statement"
+    = SELECT_TOKEN _ "{" _ cases:CommClause* _ "}" EOS {
+        return { type: "SelectStatement", cases }
+    }
```
Comment (shenyih0ng) [inline src/go-slang/parser/go.pegjs]:
```suggestion
    = SELECT_TOKEN _ "{" _ cases:CommClause* _ "}" EOS {
        return { type: "SelectStatement", cases }
      }
```

**Review by shenyih0ng — APPROVED** [review]
Let's add in the generated parser code (`go.js`) as well before merging. Not sure if we want to add in the corresponding ECE types (`src/go-slang/types.ts`) for the new grammars right now? I have no strong opinion on this, so imma leave to you to decide.

Otherwise, LGTM 🚀 Thanks for working on this!

File: src/go-slang/parser/go.pegjs [hunk 4]
```diff
@@ -185,6 +189,46 @@ CallExpression
         return makeNode({ type: "CallExpression", callee, args: args ?? [] })
       }
 
+/* Select Statement */
+
+SelectStatement "select statement"
+    = SELECT_TOKEN _ "{" _ cases:CommClause* _ "}" EOS {
+        return { type: "SelectStatement", cases }
```
Comment (shenyih0ng) [inline src/go-slang/parser/go.pegjs]: Also, we can wrap every node creation with `makeNode` to inject UID and location information which will come in handy for ECE implementation later on

File: src/go-slang/parser/go.pegjs (hunk 2 above)
Comment (alvynben) [reply src/go-slang/parser/go.pegjs]: Done.

File: src/go-slang/parser/go.pegjs (hunk 3 above)
Comment (alvynben) [reply src/go-slang/parser/go.pegjs]: Done.

File: src/go-slang/parser/go.pegjs (hunk 4 above)
Comment (alvynben) [reply src/go-slang/parser/go.pegjs]: Done.

File: src/go-slang/parser/go.pegjs (hunk 1 above)
Comment (alvynben) [reply src/go-slang/parser/go.pegjs]: Need some help with this, thanks.

**Pull request merged**